### PR TITLE
[common-library] global for security context

### PR DESCRIPTION
**Pull request by paologallinaharbur**
#### Is this a new chart
no

#### What this PR does / why we need it:
Add to globals and library the security context

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] Chart Version bumped
- [ ] Variables are documented in the README.md
- [x] Title of the PR starts with chart name (e.g. `[mychartname]`)

### CODE REVIEW

File: library/CHART-TEMPLATE/templates/deployment.yaml
Comment (gsanchezgavier) [inline library/CHART-TEMPLATE/templates/deployment.yaml:28]: would make sense to be possible to add the `securityContext:` key inside the helper ? so in the actual template you just need to add the helper entry?

**Review by gsanchezgavier — APPROVED** [review]
LGTM

File: library/common-library/templates/_security-context.tpl
Comment (gsanchezgavier) [reply library/common-library/templates/_security-context.tpl:3]: 💥

Comment (kang-makes) [inline library/common-library/templates/_security-context.tpl:3]: Here

File: library/common-library/templates/_security-context.tpl
Comment (kang-makes) [inline library/common-library/templates/_security-context.tpl:16]: And here.

**Review by kang-makes — COMMENTED** [review]
Same as we talked. We cannot trust that `global` exists.

**Comment by kang-makes** [comment]
I didn't refresh and I didn't see that you reviewed yourself...

I would like @roobre to review this because this is an issue that we talked about and I don't know if he is going to be happy about this.

I am owning this branch from now it you wish :)

File: library/common-library/templates/_security-context.tpl
Comment (roobre) [inline library/common-library/templates/_security-context.tpl:2]: I like the idea of `.container` and `.pod` but I think the standard goes more towards the `podSecurityContext` and `containerSecurityContext` direction ([e.g.](https://github.com/kubernetes/ingress-nginx/blob/fb72fcd81772fb3eca923897aec2d92fa5bdff41/charts/ingress-nginx/values.yaml#L811)).

Do we have a strong reason for breaking the "standard"?

Comment (kang-makes) [reply library/common-library/templates/_security-context.tpl:2]: No. I'm on it.

File: library/common-library/templates/_security-context.tpl
Comment (paologallinaharbur) [inline library/common-library/templates/_security-context.tpl:4]: logic is currently bit cumbersome due to the nested cases. I lend it of over to @kang-makes to be improved 😄

File: library/common-library/templates/_security-context.tpl
Comment (paologallinaharbur) [reply library/common-library/templates/_security-context.tpl:2]: the variable were not the same, it was just the function definition named like that to group them, but yeah better to stick to standards

File: library/CHART-TEMPLATE/templates/deployment.yaml
Comment (roobre) [inline library/CHART-TEMPLATE/templates/deployment.yaml:27]: Something thorny here is what will we do with the `privileged` flag, as it _needs_ to override some things in the security context (e.g. the `securityContext.privileged` flag). However I'm not sure if we want to discuss that in _this_ PR.

File: library/common-library/templates/_security-context.tpl
Comment (kang-makes) [reply library/common-library/templates/_security-context.tpl]: Solved in [#f27f4e9](https://github.com/newrelic/helm-charts/pull/725/commits/f27f4e99b33395141d632585fcd4c869f41c9ae1)

File: library/CHART-TEMPLATE/templates/deployment.yaml
Comment (kang-makes) [reply library/CHART-TEMPLATE/templates/deployment.yaml:27]: We could talk about defaults changing if defaults "depending-on-things" in the `nri-kubernetes` chart.

**Comment by kang-makes** [comment]
We [talked about replacing instead of merging](https://github.com/newrelic/helm-charts/pull/735#discussion_r821469445) so I changed how `securityContext` is replaced to be the same as `affinity` and `nodeSelector`.

A re-review is required. Sorry.

File: library/common-library/templates/_security-context.tpl
Comment (gsanchezgavier) [inline library/common-library/templates/_security-context.tpl:33]: to me is a bit verbose just to avoid a nesting level below (and perhaps because I'm not used to program in tpl), but i'm fine since we have the test now 😃

Comment (kang-makes) [reply library/common-library/templates/_security-context.tpl:33]: It tests that exists, is a map, and does not fail in case of being a null (which you reminded me that there is no test for that.

So I prefer that verbosity than having to see if `sprig` does a `default` left to right or right to left...

Comment (gsanchezgavier) [reply library/common-library/templates/_security-context.tpl:33]: In this case if not a map it will no fail during installation but the parameters are not going to be added (silently), but in case we just copy the yaml under the parameter and is not a map it will fail during installation. Am I right?

**Comment by kang-makes** [comment]
As @gsanchezgavier pointed out, I was testing too much in the helper so I tested if the user put a map or not.

That means that if the user put an (incorrect) array it is silently discarded and will render without it, making the user think that it has the correct configuration when it has not.

That being said, I rebased (again) and squashed everything to have a better history.

Re-re-review is required. I am really sorry.

File: library/common-library/templates/_security-context.tpl
Comment (paologallinaharbur) [inline library/common-library/templates/_security-context.tpl:5]: How can we provide default in charts without discarding alway the global one?

Example:
 - http://github.com/newrelic/helm-charts/blob/ecf47c05c7aee6c5b04500a57c643109b36fce7a/charts/nri-metadata-injection/values.yaml#L58-L58
 - http://github.com/newrelic/helm-charts/blob/ecf47c05c7aee6c5b04500a57c643109b36fce7a/charts/nri-metadata-injection/templates/admission-webhooks/job-patch/job-patchWebhook.yaml#L47-L47`

Even with the merge I imagine that with default we would always overwrite globals

**Review by gsanchezgavier — APPROVED** [review]
LGTM , but looks that are some pending topics

File: library/common-library/templates/_security-context.tpl
Comment (roobre) [reply library/common-library/templates/_security-context.tpl:5]: This is an interesting topic. I see three solutions, and I like neither of them:

1. Provide the chart "defaults" in the template rather than in the values. So Global values will override the template defaults, and local values will override globals.
2. Make global override local for this particular case
3. Add a boolean "helper" somewhere, like `global.securityContext.override` (defaulting to `false`) that will cause the global SC to override the local one when `true`.

Of all this I think the one I dislike less is 1️⃣ .

**Review by roobre — CHANGES_REQUESTED** [review]
Flagging changes requested so we don't merge this by accident until we solve the defaults problem

File: library/common-library/templates/_security-context.tpl
Comment (paologallinaharbur) [reply library/common-library/templates/_security-context.tpl:5]: I agree, I do not like 3 since it is not easy to be applied to other similar cases in the future.

I really dislike 2 since it would be absolutely counterintuitive and difficult to document.

I would go for the 1st as well

Comment (roobre) [reply library/common-library/templates/_security-context.tpl:5]: @newrelic/coreint Any last thought on this?
I think we can go for 1️⃣ tomorrow if no one has a better approach.

Comment (gsanchezgavier) [reply library/common-library/templates/_security-context.tpl:5]: i can think of some but non less ugly than the 1

Comment (roobre) [reply library/common-library/templates/_security-context.tpl:5]: I think the helper returning defaults is needed for booleans (like `hostNetwork`) as we cannot tell if the helper is returning empty or false, but in this case I think it might be more complex. We can simply do:

```yaml
{{- if include "common.securityContext.pod" . }}
  {{- include "common.securityContext.pod" . | toYaml | nindent 2 }}
{{- else }}
  runAsUser: 1000
  runAsGroup: 3000
  fsGroup: 2000
{{- end }}
```

Which is functionally the same as the `common.securityContext.podDefaults`, but might be more readable. WDYT?

Comment (kang-makes) [reply library/common-library/templates/_security-context.tpl:5]: I think the helper would be useful because of ugly things like the discussion we have to raise again:
- [newrelic-infrastructure-v3] Change hostNetwork depending if their kind is DaemonSet or Deployment [#698 ](https://github.com/newrelic/helm-charts/pull/698)
- newrelic-infrastructure-v3: default to recreate strategy for controlPlane Deployment [#697](https://github.com/newrelic/helm-charts/pull/697)

If we need something like "change defaults depending on something external" is better to have a fallback helper that can be overridable and it can allow me to change defaults in case of other flags changing (like `privileged` which is the next PR I would have to do)

Having so many moving parts is the reason I prefer the 3️⃣. Because changing defaults are evil but we are doing it in all the Helm charts because of UX.

Comment (kang-makes) [reply library/common-library/templates/_security-context.tpl:5]: The [latest commit](https://github.com/newrelic/helm-charts/pull/725/commits/0f4c594ecf9d3fcde4d859e4c41aedbd1fa4ca87) adds the ability to set defaults using a helper function because it is easier to revert it after finishing the discussion than do it later and have to wait for approvals.

Give me your thoughts, please.

File: library/common-library/templates/_security-context.tpl
Comment (alvarocabanas) [inline library/common-library/templates/_security-context.tpl:29]: Oks, so in the values there will be all the block in podDefaults and only honoured if others not set, I like this solution. I think it's clear for the users looking at the values and covers the need we had.

File: library/common-library/templates/_security-context.tpl
Comment (marcsanmi) [inline library/common-library/templates/_security-context.tpl:15]:
```suggestion
{{- /* Allows to change defaults for container security context either statically or dynamically */ -}}
```

File: library/common-library/templates/_security-context.tpl
Comment (marcsanmi) [inline library/common-library/templates/_security-context.tpl:34]:
```suggestion
{{- /* Allows to change defaults for pod security context either statically or dynamically */ -}}
```

File: library/common-library/templates/_security-context.tpl
Comment (kang-makes) [reply library/common-library/templates/_security-context.tpl]: Don't judge me, it is being a really hard Friday.

Solved in [#35fc122](https://github.com/newrelic/helm-charts/pull/725/commits/35fc12267613cb1f15d2e067f72262d70ec5f9dc)

File: library/common-library/templates/_security-context.tpl
Comment (kang-makes) [reply library/common-library/templates/_security-context.tpl]: Don't judge me, it is being a really hard Friday.

Solved in [#35fc122](https://github.com/newrelic/helm-charts/pull/725/commits/35fc12267613cb1f15d2e067f72262d70ec5f9dc)

File: library/common-library/templates/_security-context.tpl
Comment (kang-makes) [reply library/common-library/templates/_security-context.tpl:29]: `values.yaml` should be empty for the `.global` to work but the default can change using other helpers like the `privileged` which I am waiting for this to be merged to start working on it.

**Review by roobre — APPROVED** [review]
Looking great, I really like the final approach. Good job!

**Pull request merged**
